File: test/unitTest/java/org/thoughtcrime/securesms/util/DelimiterUtilTest.java
package org.thoughtcrime.securesms.util;


import org.junit.Before;
import org.junit.Test;

import static junit.framework.Assert.assertEquals;

public class DelimiterUtilTest {

  @Before
  public void setup() {}

  @Test
  public void testEscape() {
    assertEquals(DelimiterUtil.escape("MTV Music", ' '), "MTV\\ Music");
    assertEquals(DelimiterUtil.escape("MTV  Music", ' '), "MTV\\ \\ Music");

    assertEquals(DelimiterUtil.escape("MTV,Music", ','), "MTV\\,Music");
    assertEquals(DelimiterUtil.escape("MTV,,Music", ','), "MTV\\,\\,Music");

    assertEquals(DelimiterUtil.escape("MTV Music", '+'), "MTV Music");
  }

  @Test
  public void testSplit() {
    String[] parts = DelimiterUtil.split("MTV\\ Music", ' ');
    assertEquals(parts.length, 1);
    assertEquals(parts[0], "MTV\\ Music");

    parts = DelimiterUtil.split("MTV Music", ' ');
    assertEquals(parts.length, 2);
    assertEquals(parts[0], "MTV");
    assertEquals(parts[1], "Music");
  }

  @Test
  public void testEscapeSplit() {
    String   input        = "MTV Music";
    String   intermediate = DelimiterUtil.escape(input, ' ');
    String[] parts        = DelimiterUtil.split(intermediate, ' ');

    assertEquals(parts.length, 1);
    assertEquals(parts[0], "MTV\\ Music");
    assertEquals(DelimiterUtil.unescape(parts[0], ' '), "MTV Music");

    input        = "MTV\\ Music";
    intermediate = DelimiterUtil.escape(input, ' ');
    parts        = DelimiterUtil.split(intermediate, ' ');

    assertEquals(parts.length, 1);
    assertEquals(parts[0], "MTV\\\\ Music");
    assertEquals(DelimiterUtil.unescape(parts[0], ' '), "MTV\\ Music");
  }

}

File: core/util-jvm/src/test/java/org/signal/core/util/DelimiterUtilTest.kt
package org.signal.core.util

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Test

class DelimiterUtilTest {
  @Test
  fun testEscape() {
    assertEquals("MTV\\ Music", DelimiterUtil.escape("MTV Music", ' '))
    assertEquals("MTV\\ \\ Music", DelimiterUtil.escape("MTV  Music", ' '))

    assertEquals("MTV\\,Music", DelimiterUtil.escape("MTV,Music", ','))
    assertEquals("MTV\\,\\,Music", DelimiterUtil.escape("MTV,,Music", ','))

    assertEquals("MTV Music", DelimiterUtil.escape("MTV Music", '+'))
  }

  @Test
  fun testSplit() {
    assertArrayEquals(
      arrayOf("MTV\\ Music"),
      DelimiterUtil.split("MTV\\ Music", ' ')
    )
    assertArrayEquals(
      arrayOf("MTV", "Music"),
      DelimiterUtil.split("MTV Music", ' ')
    )
  }

  @Test
  fun testEscapeSplit() {
    "MTV Music".let { input ->
      val intermediate = DelimiterUtil.escape(input, ' ')
      val parts = DelimiterUtil.split(intermediate, ' ')
      assertEquals("MTV\\ Music", parts.single())
      assertEquals("MTV Music", DelimiterUtil.unescape(parts.single(), ' '))
    }

    "MTV\\ Music".let { input ->
      val intermediate = DelimiterUtil.escape(input, ' ')
      val parts = DelimiterUtil.split(intermediate, ' ')
      assertEquals("MTV\\\\ Music", parts.single())
      assertEquals("MTV\\ Music", DelimiterUtil.unescape(parts.single(), ' '))
    }
  }
}
